src/core/broker_config.cpp

```diff
@@ -17,80 +17,282 @@
+  // The expected value type
+  const json::value_t type;
+  // A function to validate and store the value in the config's settings structure.
+  std::function<bool(const json&, BrokerConfig&)> validate_and_store;
+  // A function to store the default setting in the settings structure if the value is not present
+  // in the config file.
+  std::function<void(BrokerConfig&)> store_default;
+};
+
+// The CID must be a string representation of a UUID.
+bool ValidateAndStoreCid(const json& val, BrokerConfig& config)
+{
+  std::string str_val = val;
+  return etcpal_string_to_uuid(&config.settings.cid, str_val.c_str(), str_val.size());
+}
+
+// The UID takes the form:
+// "uid": {
+//   "type": < "static" | "dynamic" >,
+//   "manufacturer_id": <number, always present>,
+//   "device_id": <number, present only if type is "static">
+// }
+bool ValidateAndStoreUid(const json& val, BrokerConfig& config)
+{
+  if (val.contains("type") && val["type"].is_string() && val.contains("manufacturer_id") &&
+      val["manufacturer_id"].is_number_integer())
   {
-    "cid": "",
-    "uid_type": "dynamic",
-    "uid": "",
+    std::string type = val["type"];
+    int64_t manufacturer_id = val["manufacturer_id"];
 
-    "dns_sd": {
-      "service_instance_name": "",
-      "manufacturer": "ETC",
-      "model": "RDMnet Broker",
-      "scope": ""
-    },
+    if (manufacturer_id > 0 && manufacturer_id < 0x8000)
+    {
+      if (type == "static" && val.contains("device_id") && val["device_id"].is_number_integer())
+      {
+        int64_t device_id = val["device_id"];
+        if (device_id >= 0 && device_id <= 0xffffffff)
+        {
+          RdmUid static_uid;
+          static_uid.manu = static_cast<uint16_t>(manufacturer_id);
+          static_uid.id = static_cast<uint32_t>(device_id);
+          config.settings.SetStaticUid(static_uid);
+          return true;
+        }
+      }
+      else if (type == "dynamic" && !val.contains("device_id"))
+      {
+        config.settings.SetDynamicUid(static_cast<uint16_t>(manufacturer_id));
+        return true;
+      }
+    }
+  }
+  return false;
+}
 
-    "max_connections": 0,
-    "max_controllers": 0,
-    "max_controller_messages": 500,
-    "max_devices": 0,
-    "max_device_messages": 500,
-    "max_reject_connections": 1000
+// Store a generic string.
+bool ValidateAndStoreString(const json& val, std::string& string, size_t max_size, bool truncation_allowed = true)
+{
+  std::string str_val = val;
+  if (!str_val.empty())
+  {
+    if (truncation_allowed)
+    {
+      string = str_val.substr(0, max_size);
+      return true;
+    }
+    else if (str_val.length() <= max_size)
+    {
+      string = str_val;
+      return true;
+    }
   }
-)";
+  return false;
+}
 
-BrokerConfig::BrokerConfig()
+// Validate an arithmetic type and set it in the settings struct if it is within the valid range
+// for its type.
+template <class IntType>
+bool ValidateAndStoreInt(const json& val, IntType& setting)
 {
```

Comment:
static_asserts on template parameters can help give users more useful compiler errors. In this case something like `static_assert(std::is_integral<IntType>(), "IntType must be an integral type")`. I don't think it matters much in this case because this template is an implementation detail and the type name IntType is pretty obvious. It becomes more useful in other kinds of template code.

Reply:
I'll still do it for practice and maintainability.